flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/tableImpl.scala

```diff
@@ -451,6 +450,117 @@ class TableImpl(
+    }
+
+    val childFields = logicalPlan.output.map(a => UnresolvedFieldReference(a.name))
+
+    if (replaceIfExist) {
+
+      val finalFields = new ListBuffer[Expression]()
+      val addFields = fields.map(expressionBridge.bridge)
+      childFields.foreach(e => finalFields.append(e))
+
+      // replace field if exist.
+      addFields.foreach {
+        case e@Alias(_, name, _) =>
+          val index = finalFields.indexWhere(p => p match {
+            case u: UnresolvedFieldReference => u.name.equalsIgnoreCase(name)
+            case a: Alias => a.name.equalsIgnoreCase(name)
+            case _ => false
+          })
+          if (index >= 0) {
+            finalFields(index) = e
+          } else {
+            finalFields.append(e)
+          }
+        case e =>
+          throw new TableException(
+            s"Should add an alias to the [$e], if replaceIfExist is true.")
+      }
+      select(finalFields: _*)
+    } else {
+      select(childFields ++ fields:_*)
+    }
+  }
+
+  override def renameColumns(fields: String): Table = {
+    renameColumns(ExpressionParser.parseExpressionList(fields): _*)
+  }
+
+  override def renameColumns(fields: Expression*): Table = {
+
+    val childFields = logicalPlan.output.map(a => UnresolvedFieldReference(a.name))
+    val finalFields = childFields.map(e => e.asInstanceOf[Expression]).toArray
+
+    // Rename existing fields
+    fields.map(expressionBridge.bridge).foreach {
+      case e@Alias(child: UnresolvedFieldReference, _, _) =>
+        val index = finalFields.indexWhere(p => p match {
+          case u: UnresolvedFieldReference => u.name.equalsIgnoreCase(child.name)
+          case _ => false
+        })
+        if (index >= 0) {
+          finalFields(index) = e
+        } else {
+          throw new TableException(s"Rename field [${child.name}] does not exist in source table.")
+        }
+      case e =>
+        throw new TableException(
+          s"Unexpected field expression type [$e]. " +
+            s"Renaming must add an alias to the original field, e.g., a as a1.")
+    }
+    select(finalFields: _*)
+  }
+
+  override def dropColumns(fields: String): Table = {
+    dropColumns(ExpressionParser.parseExpressionList(fields): _*)
+  }
+
+  override def dropColumns(fields: Expression*): Table = {
+    val childFields = logicalPlan.output.map(a => UnresolvedFieldReference(a.name))
+    val dropFields = fields.map(expressionBridge.bridge)
+
+    val finalFields = childFields.toBuffer
+
+    // Remove the fields which should be delete in the final list
+    dropFields.foreach {
+      case UnresolvedFieldReference(name) =>
+        val index = finalFields.indexWhere(p => p match {
+          case u: UnresolvedFieldReference => u.name.equalsIgnoreCase(name)
+          case _ => false
+        })
+
+        if (index >= 0) {
+          finalFields.remove(index)
+        } else {
+          throw new TableException(s"Drop field [${name}] does not exist in source table.")
```

Comment:
Should we allow dropped fields contain the same field name multiple times? such as dropColumns("a, a")

Reply:
I think this is a bad usage, this query should throw an exception? what do you think?

Reply:
Agree. Currently it will report exception "Drop field [a] does not exist in source table." for the above case. May be we should throw a more meaningful error message?